src/hotspot/share/jfr/periodic/sampling/jfrCPUTimeThreadSampler.cpp

```diff
@@ -0,0 +1,766 @@
+
+static void set_timer_time(timer_t timerid, int64_t period_nanos) {
+  struct itimerspec its;
+  if (period_nanos == 0) {
+    its.it_interval.tv_sec = 0;
+    its.it_interval.tv_nsec = 0;
+  } else {
+    its.it_interval.tv_sec = period_nanos / NANOSECS_PER_SEC;
+    its.it_interval.tv_nsec = period_nanos % NANOSECS_PER_SEC;
+  }
+  its.it_value = its.it_interval;
+  if (timer_settime(timerid, 0, &its, nullptr) == -1) {
+    warning("Failed to set timer for thread sampling: %s", os::strerror(os::get_last_error()));
+  }
+}
+
+bool JfrCPUSamplerThread::create_timer_for_thread(JavaThread* thread, timer_t& timerid) {
+  timer_t t;
+  struct sigevent sev;
+  sev.sigev_notify = SIGEV_THREAD_ID;
+  sev.sigev_signo = SIG;
+  sev.sigev_value.sival_ptr = &t;
+  ((int*)&sev.sigev_notify)[1] = thread->osthread()->thread_id();
+  clockid_t clock;
+  int err = pthread_getcpuclockid(thread->osthread()->pthread_id(), &clock);
+  if (err != 0) {
+    log_error(jfr)("Failed to get clock for thread sampling: %s", os::strerror(err));
+    return false;
+  }
+  if (timer_create(clock, &sev, &t) < 0) {
+    return false;
+  }
+  int64_t period = get_sampling_period();
+  if (period != 0) {
+    set_timer_time(t, period);
+  }
+  timerid = t;
+  return true;
+}
+
+
+void JfrCPUSamplerThread::stop_signal_handlers() {
+  // set the stop signal bit
+  Atomic::or_then_fetch(&_active_signal_handlers, STOP_SIGNAL_BIT, memory_order_acq_rel);
+  while (Atomic::load_acquire(&_active_signal_handlers) > STOP_SIGNAL_BIT) {
+    // wait for all signal handlers to finish
+    os::naked_short_nanosleep(1000);
+  }
+}
+
+// returns false if the stop signal bit was set, true otherwise
+bool JfrCPUSamplerThread::increment_signal_handler_count() {
+  // increment the count of active signal handlers
+  u4 old_value = Atomic::fetch_then_add(&_active_signal_handlers, (u4)1, memory_order_acq_rel);
+  if ((old_value & STOP_SIGNAL_BIT) != 0) {
+    // if the stop signal bit was set, we are not allowed to increment
+    Atomic::dec(&_active_signal_handlers, memory_order_acq_rel);
+    return false;
+  }
+  return true;
+}
+
+void JfrCPUSamplerThread::decrement_signal_handler_count() {
+  Atomic::dec(&_active_signal_handlers, memory_order_acq_rel);
+}
+
+void JfrCPUSamplerThread::allow_signal_handlers() {
+  Atomic::release_store(&_active_signal_handlers, (u4)0);
+}
+
+class VM_CPUTimeSamplerThreadInitializer : public VM_Operation {
+ private:
+  JfrCPUSamplerThread *_sampler;
+ public:
+
+  VM_CPUTimeSamplerThreadInitializer(JfrCPUSamplerThread* sampler) : _sampler(sampler) {
+  }
+
+  VMOp_Type type() const { return VMOp_CPUTimeSamplerThreadInitializer; }
+  void doit() {
+    JfrJavaThreadIterator iter;
+    while (iter.has_next()) {
+      _sampler->on_javathread_create(iter.next());
+    }
+  };
+};
+
+bool JfrCPUSamplerThread::init_timers() {
+  // install sig handler for sig
+  if ((s8)PosixSignals::install_generic_signal_handler(SIG, (void*)::handle_timer_signal) == -1) {
```

Comment:
Comparing return value to `(void*)-1` would be cleaner.
But the main problem is that it only checks for `sigaction` failure (which normally never happens), however, we should also check if there was a custom signal handler set _before_ installing our own handler, i.e. old handler is not SIG_IGN or SIG_DFL or `handle_timer_signal`.